devmapper: udev sync is a requirement

vbatts opened the pull request:
Rather than disabling devmapper for static builds, this will get more at the root. Udev sync is a prerequisite for devicemapper. That sync is false on static builds as well.

- closes  https://github.com/docker/docker/issues/4036
- closes #10664

vbatts committed:
devmapper: udev sync is a requirement

closes #10664
closes #4036

Signed-off-by: Vincent Batts <vbatts@redhat.com>

GordonTheTurtle commented:
Can you please sign your commits following these rules:

https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

The easiest way to do this is to amend the last commit:

~~~console
$ git clone -b "vbatts-dm_sync_is_required" git@github.com:vbatts/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

vbatts commented:
perhaps this also needs a storage option parameter like `dm.use_regardless_of_udev_sync_i_know_what_i_am_asking_for=true`
for the folks that are using this on ubuntu and being affected, but we can't help due the version of their udev and libdevmapper...

jfrazelle commented:
we run tests on a fedora machine on every push to master we could make sure its tested there

unclejack commented:
@vbatts I think we need a special flag to allow this to be bypassed so that it can be tested.

vbatts commented:
Will do
On Mar 16, 2015 5:15 PM, "unclejack" <notifications@github.com> wrote:

> @vbatts <https://github.com/vbatts> I think we need a special flag to
> allow this to be bypassed so that it can be tested.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11412#issuecomment-81945050>.
>

vbatts committed:
devmapper: storage-opt override for udev sync

This provides an override for forcing the daemon to still attempt
running the devicemapper driver even when udev sync is not supported.

Intended to be a very clear impairment for those choosing to use it. If
udev sync is false, there will still be an error in the daemon logs,
even when the override is in place. The docs have an explicit WARNING.

Including link to the docs for users that encounter this daemon error
during an upgrade.

Signed-off-by: Vincent Batts <vbatts@redhat.com>

vbatts commented:
@jfrazelle @unclejack @rhvgoyal added a flag

rhvgoyal commented:
@vbatts 

Few more suggestions for flag name.

"dm.ignore_udev_sync_check"
"dm.override_udev_sync_check"

rhvgoyal commented:
@vbatts 

Can we also modify tests to use above flag by default?

vbatts commented:
perhaps 'override...', because we are still checking it, so we can log in the daemon that it's an issue.

rhvgoyal commented:
@vbatts 

"dm.override_udev_sync_check" sounds good.

vbatts commented:
@rhvgoyal updated PTAL

vbatts commented:
and added a default for the tests to override the sync requirement

rhvgoyal commented:
@vbatts 

This patch looks good to me. I have a general question though about tests. Generic tests in docker are run with what graphdriver?

For example, say, I want to run integration tests and one of the test is whether container creation is successful or not. Will I run same test against all graph driver or there is a default graph driver which will always be used or something else?

vbatts commented:
vfs is the driver the Docker-in-docker daemon is running for tests.

On Tue, Mar 17, 2015 at 1:35 PM, Vivek Goyal <notifications@github.com>
wrote:

> @vbatts <https://github.com/vbatts>
>
> This patch looks good to me. I have a general question though about tests.
> Generic tests in docker are run with what graphdriver?
>
> For example, say, I want to run integration tests and one of the test is
> whether container creation is successful or not. Will I run same test
> against all graph driver or there is a default graph driver which will
> always be used or something else?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11412#issuecomment-82491526>.
>

rhvgoyal commented:
Ok, thanks.

LGTM.

rhvgoyal commented:
@vbatts 

Thinking more about it. Actually this change is not upgrade proof. Those who are using device mapper and when they upgrade docker, suddenly docker daemon will stop running and that's not good.

vbatts commented:
that's why the storage-opt flag is there. It's not supported in the way
they're running it presently.
if with the upgrade, they are then required to add the flag, it will be an
explicit indication that they're running it in an unsupported way, but
still have the ability to do so. Even if for migrating to a supported
setup, or switching to a docker binary that supports udev sync.



On Tue, Mar 17, 2015 at 1:41 PM, Vivek Goyal <notifications@github.com>
wrote:

> @vbatts <https://github.com/vbatts>
>
> Thinking more about it. Actually this change is not upgrade proof. Those
> who are using device mapper and when they upgrade docker, suddenly docker
> daemon will stop running and that's not good.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11412#issuecomment-82493921>.
>

rhvgoyal commented:
Generally I am not a big fan of such approach. Typically first we move users to a different storage driver slowly before switch is made.

Given that so many people are complaining, clearly lot of people are using it. They will be upset when upgraded docker suddenly does not start.

In kernel, we have a deprecated file where we put a feature which we want to deprecate. Send loud warnings if user is using those features in an effort to migrate them to alternate means. And once we are reasonably sure that most of the user's have migrated, we make the switch.

Personally, I would rather continue to give loud warnings and tell users to move to aufs or overlayfs or move to dynamic binary when they complain, instead of breaking working setup.  Yes, flag is there, but it is very annoying to figure out that one needs to pass a new flag before one can proceed.

All I am saying that existing users might not like this change.

unclejack commented:
Perhaps calling it something like `unsafe_override_udev_sync_check` would send a very clear message that this isn't safe, nor supported. Whatever it's called, we probably want to have something like `unsafe` or `unsupported` in the flag itself to make it clear this voids the warranty.

rhvgoyal commented:
@unclejack 

I don't think that naming of flag is an issue here. Problem is that docker 1.5 was working in a setup and same setup will suddenly stop working on 1.6. And that's a problem.

We agree that whatever was working in 1.5 is not supportable, but that ship has sailed and people are already using it. I think instead of failing it completely, a better approach is to just give loud warnings that is is not supported and will go away in future, write documentation and point people to documentation.

And once we are reasonably sure that people have migrated to different solutions, then make this switch, IMHO.

unclejack commented:
@rhvgoyal You're right, but those users also need to know as soon as possible that they're exposing themselves and any potential clients they may have to data corruption/loss on the Docker daemons where they're running such a setup.

We need to provide apt and yum repositories which are distro specific in order to be able to provide dynamically linked binaires. That's a requirement for getting this merged, IMO.

vbatts commented:
should we inverse this flag for now, so that we can make it available to
the pedantic. and once the requirements are met, we can switch this default
to fail support when udev sync is not supported?

On Wed, Mar 18, 2015 at 12:11 PM, unclejack <notifications@github.com>
wrote:

> @rhvgoyal <https://github.com/rhvgoyal> You're right, but those users
> also need to know as soon as possible that they're exposing themselves and
> any potential clients they may have to data corruption/loss on the Docker
> daemons where they're running such a setup.
>
> We need to provide apt and yum repositories which are distro specific in
> order to be able to provide dynamically linked binaires. That's a
> requirement for getting this merged, IMO.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11412#issuecomment-83038585>.
>

rhvgoyal commented:
@vbatts 

By inversing the flag you mean provide a flag to fail initialization of devicemapper if udev sync is not supported. And without presence of flag, you will allow devicemapper to initialize?

If yes, I am not sure how useful it is. If somebody knows this issue then flag does not help him. And if somebody does not know the issue, then he does not know the flag either and then its of no use to him. 

I think right now we seem to have two choices only and we need to pick our poison.

- Just continue to give loud warnings and write documentation to point to users in right direction.

- Implement this flag and in error message tell user why daemon has failed and if they want to 
  override this behavior how to do it. That way, it will atleast make it easy for users to find this new
  flag. So this will still be annoying but I guess less painful. One will have to still write documentation
   to explain what are the alternatives.

vbatts commented:
@rhvgoyal will do

vbatts commented:
rebased and updated. 
PTAL

rhvgoyal commented:
LGTM

tiborvass commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -963,9 +965,11 @@ func (devices *DeviceSet) initDevmapper(doInit bool) error {
 	// https://github.com/docker/docker/issues/4036
 	if supported := devicemapper.UdevSetSyncSupport(true); !supported {
-		logrus.Warnf("Udev sync is not supported. This will lead to unexpected behavior, data loss and errors")
+		logrus.Errorf("Udev sync is not supported. This will lead to unexpected behavior, data loss and errors. For more information, see http://docs.docker.com/reference/commandline/cli/#daemon-storage-driver-option")
```
`https`

vbatts replied:
... k
Though this should force a redirect if thats the case, because this is
where google sent me ...

On Tue, Apr 7, 2015 at 1:23 PM, Tibor Vass <notifications@github.com> wrote:

> In daemon/graphdriver/devmapper/deviceset.go
> <https://github.com/docker/docker/pull/11412#discussion_r27899608>:
>
> > @@ -963,9 +965,11 @@ func (devices *DeviceSet) initDevmapper(doInit bool) error {
> >
> >  	// https://github.com/docker/docker/issues/4036
> >  	if supported := devicemapper.UdevSetSyncSupport(true); !supported {
> > -		logrus.Warnf("Udev sync is not supported. This will lead to unexpected behavior, data loss and errors")
> > +		logrus.Errorf("Udev sync is not supported. This will lead to unexpected behavior, data loss and errors. For more information, see http://docs.docker.com/reference/commandline/cli/#daemon-storage-driver-option")
>
> https
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11412/files#r27899608>.
>

vbatts commented:
fixed

tiborvass commented:
LGTM

Ping @unclejack

tiborvass commented:
@unclejack once you're good, please move to docs-review, thanks!

unclejack commented:
LGTM

tiborvass commented:
Ping @fredlf @moxiegirl @SvenDowideit

moxiegirl commented on daemon/graphdriver/devmapper/README.md:
```diff
@@ -216,3 +216,26 @@ Here is the list of supported options:
     Example use:
 
     ``docker -d --storage-opt dm.blkdiscard=false``
+
```
There were some awkward constructions that made this hard to parse and obscured the behavior.  The Notes and Warning on top of each other indicate structural issues. Hopefully, I cleared this up with some input from @tiborvass  (ty!).

Markdown: https://gist.github.com/moxiegirl/ade0634692f7423bf474

---->

*  `dm.override_udev_sync_check`

Overrides the kernel's `udev` synchronization checks. `udev` is the device
manager for the Linux kernel. By default, this value is `false` and the kernel
does not support override `udev` synchronization checks.  

Under the `false` setting, race conditions occur between the`devicemapper` and `udev` during create and cleanup. The race conditions result in errors and the
`devicemapper` fails. (For information on these failures, see
[docker#4036](https://github.com/docker/docker/issues/4036))

To allow the `docker` daemon to function properly with the `devicemapper`
driver, set `dm.override_udev_sync_check` to true:  

	$ docker -d --storage-opt dm.override_udev_sync_check=true
	
When this value is `true`, the  `devicemapper` continues and simply warns you the errors are happening 

**Note**: The ideal is to pursue a `docker` daemon and environment that does
support synchronizing with Udev. For further discussion on this topic, see 
[docker#4036](https://github.com/docker/docker/issues/4036).  Otherwise, set this flag for migrating existing daemons to a daemon
with a supported environment.

moxiegirl commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -370,6 +370,28 @@ Currently supported options are:
 
         $ docker -d --storage-opt dm.blkdiscard=false
 
+ *  `dm.override_udev_sync_check`
```
Same as above.

vbatts commented:
@moxiegirl I will update the wording. I see that it is not clear enough. :-)

vbatts commented:
docs updated PTAL

moxiegirl commented:
@vbatts Thanks for the update. You missed a few ticks around `udev` and `devicemapper` line 381 --- but I can clean those up later.  LGTM

moxiegirl commented:
ping @fredlf @SvenDowideit @jamtur01

jamtur01 commented on daemon/graphdriver/devmapper/README.md:
```diff
@@ -216,3 +216,39 @@ Here is the list of supported options:
+
+ *  `dm.override_udev_sync_check`
+
+    Overrides the `udev` synchronization checks between devicemapper and udev.
```
\`devicemapper\` \`udev\`

jamtur01 commented on daemon/graphdriver/devmapper/README.md:
```diff
@@ -216,3 +216,39 @@ Here is the list of supported options:
+    the activation and deactivation of devices for containers.
+
+    When `udev` sync is `false`, a race conditions occurs between
+    the`devicemapper` and `udev` during create and cleanup. The race conditions
```
condition

jamtur01 commented on daemon/graphdriver/devmapper/README.md:
```diff
@@ -216,3 +216,39 @@ Here is the list of supported options:
+
+    When `udev` sync is `false`, a race conditions occurs between
+    the`devicemapper` and `udev` during create and cleanup. The race conditions
+    result in errors and failures. (For information on these failures, see
```
results

jamtur01 commented on daemon/graphdriver/devmapper/README.md:
```diff
@@ -216,3 +216,39 @@ Here is the list of supported options:
+    When this value is `true`, the  `devicemapper` continues and simply warns
+    you the errors are happening.
+
+    **Note**: The ideal is to pursue a `docker` daemon and environment that
```
Is this formatted as a proper note?

moxiegirl replied:
No it isn't. That is my fault. @vbatts I hate to pepper ya with this formatting. If you have time to update. I've fixed the gist:

https://gist.github.com/moxiegirl/ade0634692f7423bf474

vbatts replied:
ah. let me get that.

On Fri, Apr 10, 2015 at 3:02 PM, moxiegirl <notifications@github.com> wrote:

> In daemon/graphdriver/devmapper/README.md
> <https://github.com/docker/docker/pull/11412#discussion_r28171493>:
>
> > +    the activation and deactivation of devices for containers.
> > +
> > +    When `udev` sync is `false`, a race conditions occurs between
> > +    the`devicemapper` and `udev` during create and cleanup. The race conditions
> > +    result in errors and failures. (For information on these failures, see
> > +    [docker#4036](https://github.com/docker/docker/issues/4036))
> > +
> > +    To allow the `docker` daemon to start, regardless of `udev` sync not being
> > +    supported, set `dm.override_udev_sync_check` to true:
> > +
> > +        $ docker -d --storage-opt dm.override_udev_sync_check=true
> > +
> > +    When this value is `true`, the  `devicemapper` continues and simply warns
> > +    you the errors are happening.
> > +
> > +    **Note**: The ideal is to pursue a `docker` daemon and environment that
>
> No it isn't. That is my fault. @vbatts <https://github.com/vbatts> I hate
> to pepper ya with this formatting. If you have time to update. I've fixed
> the gist:
>
> https://gist.github.com/moxiegirl/ade0634692f7423bf474
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11412/files#r28171493>.
>

jamtur01 commented on daemon/graphdriver/devmapper/README.md:
```diff
@@ -216,3 +216,39 @@ Here is the list of supported options:
+    you the errors are happening.
+
+    **Note**: The ideal is to pursue a `docker` daemon and environment that
+    does support synchronizing with Udev. For further discussion on this topic,
```
Consistency of use - udev/Udev.

jamtur01 commented on daemon/graphdriver/devmapper/README.md:
```diff
@@ -216,3 +216,39 @@ Here is the list of supported options:
+    **Note**: The ideal is to pursue a `docker` daemon and environment that
+    does support synchronizing with Udev. For further discussion on this topic,
+    see [docker#4036](https://github.com/docker/docker/issues/4036).
+    Otherwise, set this flag for migrating existing daemons to a daemon with a
```
migrating existing Docker daemons ...

jamtur01 commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -370,6 +370,41 @@ Currently supported options are:
 
+ *  `dm.override_udev_sync_check`
+
+    Overrides the `udev` synchronization checks between devicemapper and udev.
```
Comments as above.

jamtur01 commented:
@vbatts Did you push an update? Perhaps wait until I get a chance to finish the review? :) Makes it hard to go back and see what's changed.

vbatts commented:
@jamtur01 sorry! i thought i saw them slow down, so i pushed :-\

jamtur01 commented:
No worries - love the lightning fast fixes! :)

tiborvass commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -370,6 +370,41 @@ Currently supported options are:
+    When `udev` sync is supported, then devicemapper and udev can coordinate
+    the activation and deactivation of devices for containers.
+
+    When `udev` sync is `false`, a race conditions occurs between
```
`udev` sync is not supported 

or

`udev` sync support is `false`

Ping @moxiegirl @jamtur01

vbatts replied:
ah, perhaps "supported" would be good there.

vbatts commented:
Notes formatting updated. PTAL

moxiegirl commented:
@jamtur01 ping back...let us know if you are satisfied. :D

jamtur01 commented:
Docs LGTM
